### ISSUE TITLE
New -bset and -bnset operators for simplified and more descriptive flag checking

### ISSUE THREAD

**Issue by KirkMunro** [PowerShell/PowerShell#10888]
# Summary of the new feature/enhancement

As a scripter,
I want `-bset` and `-bnset` operators that allow me to test if one or more bits are set or not set
so that I can use more descriptive and simplified logic when I need to test bitwise flag state.

# Proposed technical implementation details (optional)

I'm logging this issue because every time I come back to it, it bothers me.

Let's say you need to make sure one or more flags are enabled or disabled in a system that you are managing via automation. For example, consider the SSL flags on a website in IIS. You might write some logic that looks like this to make sure the appropriate flags are enabled:

```powershell
$webconfig = Get-WebConfiguration -Location $websiteName -Filter system.webserver/security/access
$sslFlagValue = $webconfig.Attributes.where{$_.Name -eq 'sslFlags'}.Value
$requiredFlags = 104
if (($sslFlagValue -band $requiredFlags) -eq $requiredFlags) {
    Set-WebConfiguration -Location $websiteName -Filter system.webserver/security/access -Value ($sslFlagValue -bor $requiredFlags)
}
```

That works, but the syntax used to check if bitwise flags are set is annoyingly awkward and cumbersome. If you're only testing one bitwise flag, you can simply use the `-band` operator, and let the result be converted to `bool`; however, if you are testing multiple bitwise flags at the same time, you need to use `-band` to separate out the flags you are testing, wrap that in parentheses, and then compare the result of that against the flags you are testing. If you have a match, all flags are set. If not, you're missing a flag.

Instead, I'd like to replace this:

`if (($something -band $flags) -eq $flags) {...}`

with this:

`if ($something -bset $flags) {...}`

It's a little syntax sugar to simplify the logic that is required when you want to check to see if one or more flags are set.

**Comment by vexx32** [comment]
Isn't this vaialble as `$something.HasFlag($flags)`? 🙂

**Comment by iSazonov** [comment]
> Isn't this available as $something.HasFlag($flags)? 

Yes, it is first question - how frequently do we manipulate enum flags?

Proposed names is cryptic. -bset implies "set", not "check/test" flag. I'd expect that the names is derived  from "the enum has flags", ex.: -flags, -hasflags, -enumflags, -enumhasflag

What is preffered syntax:
```
$enumVar -flags Flag1,Flag2,Flag3
$enumVar -flags Flag1+Flag2+Flag3
$enumVar -flags [EnumType]::Flag1,[EnumType]::Flag2,[EnumType]::Flag3
$enumVar -flags [EnumType]::Flag1+[EnumType]::Flag2+[EnumType]::Flag3
```

**Comment by KirkMunro** [comment]
> Isn't this available as `$something.HasFlag($flags)`? 🙂

You're assuming that you actually have an enumeration to work with. Sometimes you don't. In the case of IIS and the example I provided, I don't have the `HasFlag` method available to me.

> Proposed names is cryptic. -bset implies "set", not "check/test" flag. I'd expect that the names is derived from "the enum has flags", ex.: -flags, -hasflags, -enumflags, -enumhasflag

PowerShell already uses many short operator names, and since this is a bitwise operation it made sense (to me at least) to keep the same prefix as `-band`, `-bor`, and `-bnot`. In the context of bitwise operations, `-bor` is already "set". The proposal of `-bset` was meant to represent "bit is set". I also considered `-btest`, but when looking at the negative check (if a bit is not set), I didn't like `-bntest` as much as I liked `-bnset`.

Ultimately I would like to keep these short, without the word "enum" in them (because they are not necessarily enumerations), consistent with other bitwise operators, and with both positive and negative test variants. With those goals in mind I still prefer `-bset` and `-bnset` over the other proposals you suggested @iSazonov, but I'm not attached to those operator names personally.

**Comment by KirkMunro** [comment]
> Yes, it is first question - how frequently do we manipulate enum flags?

Personally, not very frequently. Yet we have `-band`, `-bor`, `-bnot`, and `-bxor` operators to work with flags because ultimately, it comes up in programming. And when it does, if you don't have an enumerated type to work with or if you don't realize that there is a `HasFlags` method available to you when you do have an enumerated type to work with, testing if flags are set (or not set) is a pain.

**Comment by SeeminglyScience** [comment]
> Yet we have `-band`, `-bor`, `-bnot`, and `-bxor` operators to work with flags

Those make sense to me because without them, you just can't do that operation.  Even as someone who uses them *very* frequently, I have a hard time seeing the benefit in adding syntactic sugar for it directly in the language.

**Comment by KirkMunro** [comment]
> Those make sense to me because without them, you just can't do that operation. Even as someone who uses them _very_ frequently, I have a hard time seeing the benefit in adding syntactic sugar for it directly in the language.

This is more than syntactic sugar.

Let me share a short, non-contrived transcript that demonstrates exactly why I think the combination of the current set of bitwise operators and the `HasFlag` enumeration method is insufficient for general scripters when they need to work with flags.

```none
# Create a flag
PS C:\> $x = [System.Management.Automation.CommandTypes]'Cmdlet,Alias'

# Get the full type name of the flag
PS C:\> $x.GetType().FullName
System.Management.Automation.CommandTypes

# Use the HasFlag method to check if multiple flags are set on the flag
PS C:\> $x.HasFlag('Cmdlet,Function')
Cannot convert argument "flag", with value: "Cmdlet,Function", for "HasFlag" to type "System.Enum": "Cannot convert value "Cmdlet,Function" to type "System.Enum". Error: "Late bound operations cannot be performed on types or methods for which ContainsGenericParameters is true.""
At line:1 char:1
+ $x.HasFlag('Cmdlet,Function')
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [], MethodException
+ FullyQualifiedErrorId : MethodArgumentConversionInvalidCastArgument

# Since that didn't work, look at HasFlag to see why
PS C:\> $x.HasFlag

OverloadDefinitions
-------------------
bool HasFlag(System.Enum flag)

# Peachy, it isn't a generic method, so you can't just use string enumeration
# values in PowerShell like you can in most other places. :|

# Ok, let's just use the bitwise operators to check if our flags are set
PS C:\> $x -band 'Cmdlet,Function'
Cmdlet

# Ok, I see that one of them is set. Now let's try that in an if statement
PS C:\> if ($x -band 'Cmdlet,Function') {'They are set'}
They are set

# Well, that's not sufficient because that tells me they're set, but really only
# one of them is set, so I really should check if the result is the set of
# enumerations that I need to test
PS C:\> if ($x -band 'Cmdlet,Function' -eq 'Cmdlet,Function') {'They are set'}
They are set

# Ok, so what's going on now? It says they are set, but they're clearly not.

# Let's look more closely
PS C:\> $x -band 'Cmdlet,Function' -eq 'Cmdlet,Function'
Alias

# Huh?
# This looks like I just identified the flags that are not set.

# I suppose I should test to see if the flags are not set, since that's really
# what I want to know
PS C:\> if ($x -band 'Cmdlet,Function' -ne 'Cmdlet,Function') {'They are not set'}

# Looking more closely again
PS C:\> $x -band 'Cmdlet,Function' -ne 'Cmdlet,Function'
0

# Now it looks like none of the flags are set.

# At some point in the commands above the scripter may realize that
# their -eq and -ne operators appear to be returning non-boolean values,
# and discover that they have a higher precedence than the bitwise
# operators, realizing they need to use parentheses.

PS C:\> if (($x -band 'Cmdlet,Function') -ne 'Cmdlet,Function') {'They are not set'}
They are not set

PS C:\> if (($x -band 'Cmdlet,Function') -eq 'Cmdlet,Function') {'They are set'}

PS C:\>
```

# It still seems like they are set, but they are not.



PS C:\> ($x -band 'Cmdlet,Function') -eq 'Cmdlet,Function'
True
PS C:\>
```

This is a realistic scenario that a scripter might go through when working with bitwise flags in PowerShell. We focus on so many things to make scripting easier, but these operators are definitely not the easiest to work with. It would be much easier and less error prone (especially since certain scenarios demonstrated above show how easy it could be to have conditional tests pass incorrectly) if we had:

1. A conditional operator that returns true if a collection of bits are set in a flag.
1. A conditional operator that returns true if a collection of bits are not set in a flag.

That's my opinion on this, but I fully respect if others disagree.

**Comment by SeeminglyScience** [comment]
I should have phrased that differently - I definitely see the potential benefit, but it's going to be pretty rare for a scripter to require exhaustive checks of bit flags.  Personally I do not think the impact would be broad enough to warrant additional operators.

Side note, `HasFlag` does work as long you explicitly type the string, e.g. `$flags.HasFlag([Reflection.BindingFlags]'Public, Instance')` (which is my preference to do in general as I find the string by itself kind of hard to follow).  Also I almost never use multiple flags on the RHS, this type of pattern seems to work well:

```powershell
if (($flags -band 'Instance') -ne 0 -and ($flags -band 'NonPublic') -ne 0)
{
}
```
